### PR TITLE
Minimum flush interval added

### LINKED ISSUE #9: Allow minimum flush interval (time based)
If the event frequency is low (e.g. during night time), the compress_spooler may suffer from very long flush intervals. There should be a possibility to config a minimum flush interval.

### PR DESCRIPTION
**Pull request by breml**
Implements a naive aproach to allow a minimum flush interval (best effort).
The queue is flushed, when the next event is processed, after the minimum
flush interval is exeeded.

Fixes #9

### CODE REVIEW

**Comment by breml** [comment]
What are the essential points to satisfy the automatic CLA test? I signed the Elasticsearch CLA this summer. Name, e-mail and Github-user are matching.

**Comment by breml** [comment]
ping?

**Comment by purbon** [comment]
@breml fix now, there was a glitch in the automatic check it got already fix. Thanks for your contribution and sorry for taking long time to come back here. Will review this soon.

**Comment by purbon** [comment]
@breml now that I see your code, to move forward It would be required that you add test for this change. We enforce this in order to have a better tests harness so we :arrow_up: our overall quality. Happy to assist you in case you need :+1:, let me know.

**Comment by breml** [comment]
@purbon added some tests. ok?

**Comment by breml** [comment]
@purbon Could you please give me some feedback to the tests?

**Comment by purbon** [comment]
please jenkins, test this.

File: .gitignore [hunk 1]
```diff
@@ -1,3 +1,4 @@
 *.gem
 Gemfile.lock
 .bundle
+*.swp
```
Comment (purbon) [inline .gitignore]: small question, why is this necessary?

File: lib/logstash/codecs/compress_spooler.rb [hunk 2]
```diff
@@ -6,12 +6,18 @@ class LogStash::Codecs::CompressSpooler < LogStash::Codecs::Base
   config :spool_size, :validate => :number, :default => 50
   config :compress_level, :validate => :number, :default => 6
 
+  # The amount of time in seconds since last flush before a flush is forced,
+  # on the next event.
+  # Values smaller than 0 disables time based flushing.
+  config :min_flush_time, :validate => :number, :default => 0
```
Comment (purbon) [inline lib/logstash/codecs/compress_spooler.rb]: as I understand, this changed behaviour, can we make it ```-1``` as default? this will make the setting backwards compatible, right? what do you think?

File: lib/logstash/codecs/compress_spooler.rb [hunk 3]
```diff
@@ -25,9 +31,10 @@ def encode(event)
     # MessagePack#pack which relies on pure Ruby object recognition
     @buffer << LogStash::Util.normalize(event.to_hash).merge(LogStash::Event::TIMESTAMP => event.timestamp.to_iso8601)
     # If necessary, we flush the buffer and get the data compressed
-    if @buffer.length >= @spool_size
+    if @buffer.length >= @spool_size || (@min_flush_time > 0 && (@last_flush + @min_flush_time < Time.now.to_i))
```
Comment (purbon) [inline lib/logstash/codecs/compress_spooler.rb]: what about wrapping this ```(@min_flush_time > 0 && (@last_flush + @min_flush_time < Time.now.to_i)``` condition in side a method? to we improve a bit readability, what do you think?

File: spec/codecs/compress_spooler_spec.rb [hunk 4]
```diff
@@ -143,6 +143,33 @@
           expect(buffer.size).to eq(2)
         end
       end
+
+      context "message spooling if min flush time is set" do
+        let(:spool_size) { 10 }
+        min_flush_time = 2
```
Comment (purbon) [inline spec/codecs/compress_spooler_spec.rb]: can we make this var also inside a let? is discouraged to have this kind of statements in rspec.

File: spec/codecs/compress_spooler_spec.rb [hunk 5]
```diff
@@ -143,6 +143,33 @@
...
+        let(:min_flush_time) { min_flush_time }
+        subject(:codec) { LogStash::Codecs::CompressSpooler.new("spool_size" => spool_size, "min_flush_time" => min_flush_time) }
```
Comment (purbon) [inline spec/codecs/compress_spooler_spec.rb]: I would recommend moving the constructor parameters to their own let, so we could make this line smaller. Also using ```described_class``` (see https://www.relishapp.com/rspec/rspec-core/docs/metadata/described-class) will help on redability.

**Comment by purbon** [comment]
@breml I left you a couple of comments to improve test, but in general are good.

File: .gitignore (hunk 1 above)
Comment (breml) [inline .gitignore]: I'm often working with vi and I would like to prevent from committing some swp files.

Comment (purbon) [inline .gitignore]: works for me.

File: lib/logstash/codecs/compress_spooler.rb (hunk 2 above)
Comment (breml) [inline lib/logstash/codecs/compress_spooler.rb]: No, the default value of 0 does not change behavior. Setting min_flush_time to 0 disables time based flushing as the second part of the condition (after ||) on line 34 will never be true, because `@min_flush_time > 0 == false`. This means, the buffer is flushed, `if @buffer.length > @spool_size` is true. This is the old behavior.

File: lib/logstash/codecs/compress_spooler.rb (hunk 3 above)
Comment (breml) [inline lib/logstash/codecs/compress_spooler.rb]: Yeah sure, I update the PR.

File: spec/codecs/compress_spooler_spec.rb (hunk 4 above)
Comment (breml) [inline spec/codecs/compress_spooler_spec.rb]: Changed this as well (could not remember why I did this in the first place).

File: spec/codecs/compress_spooler_spec.rb (hunk 5 above)
Comment (breml) [inline spec/codecs/compress_spooler_spec.rb]: OK.

**Comment by breml** [comment]
@purbon Added a new commit on Friday which implements you recommendations.

**Comment by breml** [comment]
@purbon Any feedback on this?

**Comment by breml** [comment]
@purbon, @logstash-core ping

**Comment by jsvd** [comment]
merged in a9d4ae80ad6baff40a8059ade27af238a442d6d5
